<launch>
	<!-- ARGUMENTS -->
	<arg name="use_safety" default="true"/>
	<arg name="use_localization" default="true"/>
	<arg name="use_initial_pose" default="false"/>
	<arg name="enable_control" default="true"/>
	<arg name="use_log" default="false"/>
	<arg name="unlink_sensor_link" default="false"/>
	<arg name="use_simulator" default="true" />
	<arg name="use_trajectory_tracker" default="false"/>
	<arg name="save_data" default="false"/>
	<arg name="local_planner" default="frene_planner"/> 
	<arg name="catkin_ws_folder" default="/home/$(env USER)/crest_catkin"/>
	

	<!--	<arg name="front_back_baseline" default="2.7"/>   -->

	<!-- Trajectory Tracker -->
	<arg name="max_vel" default="0.1"  />
	<arg name="max_angvel" default="4.9"  />
	<arg name="max_acc" default="3.0"  />
	<arg name="max_angacc" default="5.0"  />
	<arg name="rotate_ang" default="0.5"  />
	<arg name="limit_vel_by_avel" default="false" />
	<arg name="trajectory_tracker_type" default="ackermann_trajectory_tracker"/>
	<arg name="trajectory_path" default="/plan"/>

	<rosparam param="footprint" ns="move_base/local_costmap">[[0.30, -0.30], [0.30, 0.30], [-0.30, 0.30], [-0.30, -0.30]]</rosparam>
	<rosparam param="footprint" ns="move_base/global_costmap/">[[0.30, -0.30], [0.30, 0.30], [-0.30, 0.30], [-0.30, -0.30]]</rosparam>
	<rosparam param="footprint" ns="limiter">[[0.30, -0.30], [0.30, 0.30], [-0.30, 0.30], [-0.30, -0.30]]</rosparam>

	<node pkg="tf" type="static_transform_publisher" name="tf" args="0.03 0 0.24 0 0 0 base_link laser 20" />
	<node pkg="tf" type="static_transform_publisher" name="slink" args="0 0 0 0 0 0 base_link sensor_link 100" unless="$(arg unlink_sensor_link)" />

<!-- Driver -->

     <group unless="$(arg use_simulator)">
          <node respawn="true" pkg="mabx_driver" type="mabx_cmd_vel" name ="mabx_cmd_vel" >
     	  	<param name="max_linear_force" value="100" type="int" />
	  </node>
	  <!-- 
     	  <node pkg="mabx_driver" type="udp_sender" name ="udp_sender"  />
	   -->
     	  <node pkg="mabx_driver" type="udp_listener" name ="udp_listener"  />
		<node pkg="tf" type="static_transform_publisher" name="stf3" args="-0.02 -0.06 0.83 0.04 0.33 1.50 sensor_link camera2_link 100"/>
		<node pkg="tf" type="static_transform_publisher" name="stf5" args="0.03 0.08 0.92 -0.02 0.67 -0.04 sensor_link camera3_link 100"/>
		<node pkg="tf" type="static_transform_publisher" name="stf4" args="-0.06 -0.09 1.15 0.13 -0.54 1.48 sensor_link camera_link 100"/>

		<arg name="port" default="/dev/actuators/tfrog" />		 
		<arg name="port_lrf" default="192.168.150.10" />

		<!-- LRF -->
	   	<node name="urg_node" pkg="urg_node" type="urg_node" unless="$(arg use_log)">
			<param name="serial_port" value="$(arg port_lrf)" type="string" />
			<param name="frame_id" type="string" value="laser" />
			<param name="angle_min" value="-2.0" />
			<param name="angle_max" value="2.0" />
			<param name="calibrate_time" type="bool" value="true" />
			<remap from="/scan" to="/scan_orig0" />
		</node>
		
		<!-- LOCALIZATION -->
		<group if="$(arg use_localization)">
			<group if="$(arg use_initial_pose)">
				<rosparam param="initial_pose_x" ns="amcl">0</rosparam>
				<rosparam param="initial_pose_y" ns="amcl">0</rosparam>
				<rosparam param="initial_pose_a" ns="amcl">0</rosparam>
			</group>
			<node name="amcl" pkg="amcl" type="amcl">
				<param name="odom_alpha1" type="double" value="4" />
				<param name="odom_alpha2" type="double" value="4" />
				<param name="odom_alpha3" type="double" value="2" />
				<param name="odom_alpha4" type="double" value="2" />
				<param name="laser_max_beams" type="int" value="200" />
				<param name="update_min_d" type="double" value="0.05" />
				<param name="update_min_a" type="double" value="0.08" />
				<param name="use_map_topic" type="bool" value="true" />
				<param name="laser_model_type" type="string" value="beam" />
				<param name="min_particles" type="int" value="30" />
				<param name="max_particles" type="int" value="100" />
				<param name="laser_likelihood_max_dist" type="double" value="0.5" />
			</node>
		</group>
	</group>

	<group if="$(arg use_simulator)">
		<group if="$(arg use_initial_pose)">
			<rosparam param="delta_x" ns="amcl">0</rosparam>
			<rosparam param="delta_y" ns="amcl">0</rosparam>
			<rosparam param="delta_yaw" ns="amcl">0</rosparam>
		</group>

		 <include file="$(find simple_simulator)/launch/ackermann.launch" >
			<arg name="use_fake_localization" value="true" />
			<arg name="remap_scan_topic" value="true" />
			<arg name="scan_topic" value="/scan_orig0" />
			<!-- <arg name="front_back_baseline" value="2.7"/> -->
			<arg name="front_back_baseline" value="0.5"/>

			 <arg name="initialPoseX" value="0.0"/>
                        <arg name="initialPoseY" value="0.0"/>
                        <arg name="initialPoseTheta" value="0.0"/>-->
		</include> 
		

	          <node name="dummy_transform" pkg="pose_state_estimator" type="dummy_transform" output="log">
			<param name="read_parent_frame_id" value="map" type="string" />
			<param name="read_child_frame_id" value="odometry_frame" type="string" />
			<param name="write_parent_frame_id" value="map" type="string" />
			<param name="write_child_frame_id" value="base_link" type="string" />
		</node>
	</group>
	
	<node name="unify" pkg="laser_preprocess" type="uni_density">
		<param name="scan" value="/scan_orig" type="string" />
		<param name="uscan" value="/scan" type="string" />
		<param name="cull_dist" value="0.05" type="double" />
		<param name="mixed_dist" value="0.15" type="double" />
	</node>
	<node pkg="laser_filters" type="scan_to_scan_filter_chain" output="screen" name="laser_filter">
		<remap from="scan" to="/scan_orig0" />
		<remap from="scan_filtered" to="/scan_orig" />
		<rosparam command="load" file="$(find ypspur_ros_bridge)/config/shadow_filter.yaml" />
	</node>


	<!-- LAUNCH SAFETY LIMITER -->
	<include file="$(find safety_limiter)/launch/safety_limiter.launch" if="$(arg use_safety)">
		<arg name="cmd_vel" value="/limiter/cmd_vel" />
		<arg name="cloud" value="/limiter/cloud" />
		<arg name="odom" value="/odom" />
		<arg name="cmd_vel_out" value="/driver/cmd_vel" /> 
		<arg name="lin_acc" value="0.6" />
		<arg name="ang_acc" value="1.0" />
		<arg name="lin_margin" value="0.15" />
		<arg name="ang_margin" value="0.1" />
		<arg name="lin_esc_margin" value="0.15" />
		<arg name="ang_esc_margin" value="0.3" />
		<arg name="unify_scan" value="true" />
	</include>

	<group if="$(arg use_trajectory_tracker)">
	<node pkg="trajectory_tracker" type="ackermann_trajectory_tracker" name="ackermann_trajectory_tracker" output="screen">
		<param name="cmd_vel" value="/cmd_vel_mux/path_controller" type="string" />
		<!-- <param name="path" value="/plan" type="string" /> -->
		<param name="path" value="$(arg trajectory_path)" type="string" />
		<param name="path_step" value="3" type="int" />
		<param name="hz" value="50" type="double" />


		<param name="max_vel"    value="$(arg max_vel)"  />
		<param name="max_steering" value="2.0"   />
		<param name="max_acc"    value="$(arg max_acc)"   />
		<param name="max_angacc" value="$(arg max_angacc)"   />
		<param name="rotate_ang" value="$(arg rotate_ang)"  />
		<param name="limit_vel_by_avel" value="$(arg limit_vel_by_avel)" />


		<param name="curv_forward" value="1.2" type="double" />
		<param name="look_forward" value="0.1" type="double" />
		<param name="k_dist" value="8" type="double" />
		<param name="k_ang" value="12" type="double" />
		<param name="k_avel" value="9" type="double" />
		<param name="dist_lim" value="0.3" type="double" />
		<param name="distance_angle_factor" value="0.5" type="double" />
		<param name="goal_tolerance_dist" value="0.3" type="double" />
		<param name="goal_tolerance_ang" value="0.15" type="double" />
		<param name="allow_backward" value="false" type="bool" />
		<param name="use_logged_velocity" value="true" type="bool" />
		<param name="low_pass_filter_coef" value="0.0" type="double" />
		<param name="max_driving_wheel_w" value="3.0" type="double" />

		<param name="dist_stop" value="2.3" type="double" />
	</node>
	</group>

	<group if="$(arg save_data)">
        <node name="data_saver" pkg="mabx_driver" type="data_saver">
		<param name="queSize" value="10" type="int" />
		<param name="samplingTimeHz" value="100" type="int" />
		<param name="vehicleSpeedTopicName" value="/vehicle_speed" type="string" />
		<param name="vehicleYawRateTopicName" value="/vehicle_yawrate" type="string" />
		<param name="vehicleAccelerationTopicName" value="/vehicle_acceleration" type="string" />
		<param name="udpSensorPacketTopicName" value="/udp_sensor_packet" type="string" />
		<param name="udpControlPacketTopicName" value="/udp_control_packet" type="string" />
		<param name="currentPoseTopicName" value="/current_pose" type="string" />
		<param name="driverCmdVelTopicName" value="/driver/cmd_vel" type="string" />
		<param name="folderName" value="$(env HOME)/DATA" type="string" />
	</node>
	</group>

	<node name="car_marker" pkg="test_tools" type="car_marker" >
		<param name="frame_id" value="/base_link" type="string" />
		<param name="x_pos" value="1.35" type="double" />
	</node>


	<!-- LOCAL PATH PLANNER -->
	<include file="$(arg catkin_ws_folder)/local_planner/$(arg local_planner).launch">
		</include>

</launch>
